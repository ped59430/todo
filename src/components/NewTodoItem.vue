<template>
  <q-item class="column">
    <div class="full-width" v-if="errorMessage !== ''">
      <span class="text-red">{{ errorMessage }}</span>
    </div>
    <div class="full-width row">
      <q-input
        class="col-xs-8 col-md-10"
        ref="textinput"
        v-model="value"
        filled
        autofocus
        @keydown.enter="submit"
      />
      <q-btn class="col-xs-4 col-md-2 flex-center" @click="submit"
        >Submit</q-btn
      >
    </div>
  </q-item>
</template>
<script>
export default {
  data() {
    return {
      value: "",
      errorMessage: ""
    };
  },
  methods: {
    async submit() {
      if (this.value === "") {
        this.errorMessage = "No data entered";
        return;
      }
      await this.addTodo({
        description: this.value,
        done: false
      });
      this.errorMessage = "";
      this.value = "";
    },
    addTodo(todo) {
      return this.$store.dispatch("todo/serverAddTodo", todo);
    }
  }
};
</script>
