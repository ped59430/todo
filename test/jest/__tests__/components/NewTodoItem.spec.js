/* eslint-disable */
/**
 * @jest-environment jsdom
 */

import { mount, createLocalVue, shallowMount } from "@vue/test-utils";
import * as All from "quasar";
import Vuex from "vuex";
import NewTodoItem from "~/src/components/NewTodoItem.vue";
const { Quasar, date } = All;

const components = Object.keys(All).reduce((object, key) => {
  const val = All[key];
  if (val && val.component && val.component.name != null) {
    object[key] = val;
  }
  return object;
}, {});

describe("NewTodoItem", () => {
  const localVue = createLocalVue();
  localVue.use(Quasar, { components });
  localVue.use(Vuex);

  let store;
  let mutations;
  let wrapper;
  let actions;

  beforeAll(() => {
    mutations = {
      MUTATION: jest.fn()
    };
    actions = {
      serverAddTodo: jest.fn()
    };
    store = new Vuex.Store({
      modules: {
        todo: {
          namespaced: true,
          state: {},
          mutations,
          actions
        }
      }
    });
    wrapper = mount(NewTodoItem, {
      store,
      localVue
    });
  });

  it("snapshot testing", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("submit method should fail if value is empty", () => {
    const { vm } = wrapper;
    vm.value = "";
    vm.submit();
    expect(vm.errorMessage).toEqual("No data entered");
  });

  it("submit successfully when there is value", async () => {
    const { vm } = wrapper;
    const description = "asdqwe";
    vm.value = description;
    await vm.submit();
    expect(actions.serverAddTodo.mock.calls).toHaveLength(1);
    expect(actions.serverAddTodo.mock.calls[0][1]).toEqual({
      description,
      done: false
    });
    expect(vm.value).toEqual("");
  });
});
