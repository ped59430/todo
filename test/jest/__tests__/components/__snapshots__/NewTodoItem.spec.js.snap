// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewTodoItem snapshot testing 1`] = `
<div class="q-item q-item-type row no-wrap column">
  <!---->
  <div class="full-width row">
    <div class="q-field row no-wrap items-start col-xs-8 col-md-10 q-input q-field--filled">
      <div class="q-field__inner relative-position col self-stretch column justify-center">
        <div tabindex="-1" class="q-field__control relative-position row no-wrap">
          <div class="q-field__control-container col relative-position row no-wrap q-anchor--skip"><input tabindex="0" type="text" class="q-field__native"></div>
        </div>
      </div>
    </div> <button tabindex="0" type="button" class="q-btn inline q-btn-item non-selectable col-xs-4 col-md-2 flex-center q-btn--standard q-btn--rectangle q-focusable q-hoverable">
      <div tabindex="-1" class="q-focus-helper"></div>
      <div class="q-btn__content text-center col items-center q-anchor--skip justify-center row">Submit</div>
    </button>
  </div>
</div>
`;
